lib/events.js

```diff
@@ -647,3 +652,84 @@ function once(emitter, name) {
     emitter.once(name, eventListener);
   });
 }
+
+const AsyncIteratorPrototype = ObjectGetPrototypeOf(
+  ObjectGetPrototypeOf(async function* () {}).prototype);
+
+function createIterResult(value, done) {
+  return { value, done };
+}
+
+function on(emitter, event) {
+  const unconsumedEvents = [];
+  const unconsumedPromises = [];
+  let error = null;
+
+  const iterator = ObjectSetPrototypeOf({
+    next() {
+      if (error) {
+        return PromiseReject(error);
+      }
+
+      const value = unconsumedEvents.shift();
+      if (value) {
+        return PromiseResolve(createIterResult(value, false));
+      }
+
+      return new Promise(function(resolve, reject) {
+        unconsumedPromises.push({ resolve, reject });
+      });
+    },
+
+    return() {
+      emitter.removeListener(event, eventHandler);
+      emitter.removeListener('error', errorHandler);
+
+      for (const promise of unconsumedPromises) {
+        promise.resolve(createIterResult(undefined, true));
+      }
+
+      return PromiseResolve(createIterResult(undefined, true));
+    },
+
+    throw(err) {
+      if (!err || !(err instanceof Error)) {
+        throw new ERR_INVALID_ARG_TYPE('EventEmitter.AsyncIterator',
+                                       'Error', err);
+      }
+      error = err;
+      emitter.removeListener(event, eventHandler);
+      emitter.removeListener('error', errorHandler);
+    },
+
+    [SymbolAsyncIterator]() {
+      return this;
+    }
+  }, AsyncIteratorPrototype);
+
+  emitter.on(event, eventHandler);
+  emitter.on('error', errorHandler);
+
+  return iterator;
+
+  function eventHandler(...args) {
+    const promise = unconsumedPromises.shift();
+    if (promise) {
+      promise.resolve(createIterResult(args, false));
+    } else {
+      unconsumedEvents.push(args);
+    }
+  }
+
+  function errorHandler(err) {
+    if (unconsumedPromises.length > 0) {
+      for (const promise of unconsumedPromises) {
+        promise.reject(err);
+      }
+    } else {
+      emitter.removeListener(event, eventHandler);
```

Comment:
since we have removed the error listener it would be nice to mark this iterator as `done`. 

That way the following code would yield

```
while (true) {
  try { 
    const r = await itr.next()
    if (r.done) break
  } catch (err) {
    // handle err
  }
}
```

To do this we have to set a boolean `finished` variable in closure scope and return an `createIterResult(undefined, true)` in `next()` ;

Of course the `next()` function should return the error if it exists, exactly once and future next() calls should return { done: true } instead of further rejections.

Although I am unsure if a error state iterator should return an endless number of rejections when calling `next()` in a while true loop or one rejection and endless number of done=true.

Reply:
I'm not convinced, I think all future `next()` call should reject with the same error.

Reply:
@mcollina For reference, here's how a builtin async iterator works:

```js
async function* test() {
    yield 1;
    throw new Error('boom')
}
var it = test()
Promise.allSettled([it.next(), it.next(), it.next()]).then(console.log)
```

![image](https://user-images.githubusercontent.com/2352663/71249724-267def80-231e-11ea-9952-bff161249e6f.png)

Reply:
After talking to @Raynos I believe we should do the async generator behavior

Reply:
Implemented, please check